### PR TITLE
Creating compiled lambda expression math class for future use

### PR DESCRIPTION
**Pull request by tmilnthorp**
For potential future use for generic quantity structs

### CODE REVIEW

**Comment by tmilnthorp** [comment]
@ZacharyPatten I agree although it ends up a bit long-winded constantly calling ```RuntmeCompiledFunctions.Add<double, double, double>.Function(left, right)```

Calling via ```CompiledLambdas.Add<double, double, double>(left, right)``` is possible if you write it as follows. It's a bit verbose inside the class, but the rest of the codebase is cleaner IMHO.

```csharp
    internal static class CompiledLambdas
    {
        internal static TResult Multiply<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            MultiplyHelper<TLeft, TRight, TResult>.Function(left, right);

        internal static TResult Divide<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            DivideHelper<TLeft, TRight, TResult>.Function(left, right);

        internal static TResult Add<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            AddHelper<TLeft, TRight, TResult>.Function(left, right);

        internal static TResult Subtract<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            SubtractHelper<TLeft, TRight, TResult>.Function(left, right);

        internal static bool Equal<TLeft, TRight>(TLeft left, TRight right) =>
            EqualHelper<TLeft, TRight>.Function(left, right);

        internal static bool NotEqual<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            NotEqualHelper<TLeft, TRight>.Function(left, right);

        internal static bool LessThan<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            LessThanHelper<TLeft, TRight>.Function(left, right);

        internal static bool GreaterThan<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            GreaterThanHelper<TLeft, TRight>.Function(left, right);

        internal static bool GreaterThanOrEqual<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            GreaterThanOrEqualHelper<TLeft, TRight>.Function(left, right);

        internal static bool LessThanOrEqual<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            LessThanOrEqualHelper<TLeft, TRight>.Function(left, right);

        internal static TResult Modulo<TLeft, TRight, TResult>(TLeft left, TRight right) =>
            ModuloHelper<TLeft, TRight, TResult>.Function(left, right);

        #region Helper Classes

        private static class MultiplyHelper<TLeft, TRight, TResult>
        {
            internal readonly static Func<TLeft, TRight, TResult> Function =
                CreateBinaryFunction<TLeft, TRight, TResult>(Expression.Multiply);
        }

        private static class DivideHelper<TLeft, TRight, TResult>
        {
            internal readonly static Func<TLeft, TRight, TResult> Function =
                CreateBinaryFunction<TLeft, TRight, TResult>(Expression.Divide);
        }

        private static class AddHelper<TLeft, TRight, TResult>
        {
            internal readonly static Func<TLeft, TRight, TResult> Function =
                CreateBinaryFunction<TLeft, TRight, TResult>(Expression.Add);
        }

        private static class SubtractHelper<TLeft, TRight, TResult>
        {
            internal readonly static Func<TLeft, TRight, TResult> Function =
                CreateBinaryFunction<TLeft, TRight, TResult>(Expression.Subtract);
        }

        private static class EqualHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.Equal);
        }

        private static class NotEqualHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.NotEqual);
        }

        private static class LessThanHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.LessThan);
        }

        private static class GreaterThanHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.GreaterThan);
        }

        private static class GreaterThanOrEqualHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.GreaterThanOrEqual);
        }

        private static class LessThanOrEqualHelper<TLeft, TRight>
        {
            internal readonly static Func<TLeft, TRight, bool> Function =
                CreateBinaryFunction<TLeft, TRight, bool>(Expression.LessThanOrEqual);
        }

        private static class ModuloHelper<TLeft, TRight, TResult>
        {
            internal readonly static Func<TLeft, TRight, TResult> Function =
                CreateBinaryFunction<TLeft, TRight, TResult>(Expression.Modulo);
        }

        #endregion

        private static Func<TLeft, TRight, TResult> CreateBinaryFunction<TLeft, TRight, TResult>(Func<Expression, Expression, BinaryExpression> expressionCreationFunction)
        {
            var leftParameter = Expression.Parameter(typeof(TLeft), "left");
            var rightParameter = Expression.Parameter(typeof(TRight), "right");

            var binaryExpression = expressionCreationFunction(leftParameter, rightParameter);
            var lambda = Expression.Lambda<Func<TLeft, TRight, TResult>>(binaryExpression, leftParameter, rightParameter);

            return lambda.Compile();
        }
    }
```

You could even add this so you can deduce the parameters and just call ```CompiledLambdas.Add(left, right)``` for calls where left/right/result are the same types:

```csharp
        internal static T Add<T>(T left, T right) =>
            AddHelper<T, T, T>.Function(left, right);
```

**Comment by ZacharyPatten** [comment]
The C# compiler can infer generic parameters on methods. If you want to simplify the syntax, you can wrap the calls in methods and provide overloads in which the generic parameters can be inferred by the compiler.

That is what I did in my Compute class here: https://github.com/ZacharyPatten/Towel/blob/master/Sources/Towel/Mathematics/Compute.cs

You can call it like so:
```csharp
Compute.Add(1, 2);
Compute.Add(1.5f, 2.5f);
```

The compiler will never be able to infer generic parameters that are used for the return type though.

**Comment by tmilnthorp** [comment]
@ZacharyPatten yup - just looked at your example and it's exactly what I did in the above example. I'll update the code.

**Comment by angularsen** [comment]
I've only glance through this thread so far, let me know when you want a review @tmilnthorp .

**Comment by tmilnthorp** [comment]
I think it's ready. Just not used yet :)

File: UnitsNet.Tests/CompiledLambdasTests.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        [Theory]
+        [InlineData(3.0, 3.0, true)]
+        [InlineData(3.0, 2.0, false)]
+        public void EqualReturnsExpectedValues<TLeft, TRight>(TLeft left, TRight right, bool expected)
+        {
+            Assert.Equal(expected, CompiledLambdas.Equal(left, right));
+        }
+
+        [Theory]
+        [InlineData(3.0, 3.0, false)]
```
Comment (angularsen) [inline UnitsNet.Tests/CompiledLambdasTests.cs]: Nice tests

File: UnitsNet/CompiledLambdas.cs [hunk 2]
```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.Linq.Expressions;
+
+namespace UnitsNet
+{
+    /// <summary>
+    /// Compiled lambda expressions that can be invoked with generic run-time parameters.
+    /// </summary>
```
Comment (angularsen) [inline UnitsNet/CompiledLambdas.cs]: We could mention that this is all about performance vs reflection-based techniques.

File: UnitsNet/CompiledLambdas.cs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            DivideImplementation<TLeft, TRight, TResult>.Invoke(left, right);
+
+        /// <summary>
+        /// Adds the given values.
+        /// </summary>
+        /// <typeparam name="T">The type of the operation (left hand side, right hand side, and result).</typeparam>
+        /// <param name="left">The left hand side parameter.</param>
+        /// <param name="right">The right hand side parameter.</param>
+        /// <returns>The added result.</returns>
+        internal static T Add<T>(T left, T right) => Add<T, T, T>(left, right);
```
Comment (angularsen) [inline UnitsNet/CompiledLambdas.cs]: Any reason to not call `AddImplementation` directly here? Same for other places. 
The JIT probably inlines this, but I don't know if it does.

**Comment by angularsen** [comment]
Bad bot, reopening this one.

File: UnitsNet.Tests/CompiledLambdasTests.cs (hunk 1 above)
Comment (tmilnthorp) [reply UnitsNet.Tests/CompiledLambdasTests.cs:52]: 👍

File: UnitsNet/CompiledLambdas.cs (hunk 3 above)
Comment (tmilnthorp) [reply UnitsNet/CompiledLambdas.cs]: I would think so, but I called them directly instead

**Comment by angularsen** [comment]
I've forgotten a lot of the details here, but it seems I was happy the last time I reviewed it and you have now addressed my comment, so.. I'll merge this then :-)

**Comment by angularsen** [comment]
[Release UnitsNet/4.61.0 · angularsen/UnitsNet](https://github.com/angularsen/UnitsNet/releases/tag/UnitsNet%2F4.61.0)

**Pull request merged**